Data Analytics, we often use Model Development to help us predict future observations from the data we have.

A Model will help us understand the exact relationship between different variables and how these variables are used to predict the result.

First import pd, np, plt libraries 

In [1]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt

Next import the data

In [2]:
path = 'https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DA0101EN/automobileEDA.csv'


In [3]:
cars = pd.read_csv(path)
cars.head()

symboling  normalized-losses         make aspiration num-of-doors  \
0          3                122  alfa-romero        std          two   
1          3                122  alfa-romero        std          two   
2          1                122  alfa-romero        std          two   
3          2                164         audi        std         four   
4          2                164         audi        std         four   

    body-style drive-wheels engine-location  wheel-base    length  ...  \
0  convertible          rwd           front        88.6  0.811148  ...   
1  convertible          rwd           front        88.6  0.811148  ...   
2    hatchback          rwd           front        94.5  0.822681  ...   
3        sedan          fwd           front        99.8  0.848630  ...   
4        sedan          4wd           front        99.4  0.848630  ...   

   compression-ratio  horsepower  peak-rpm city-mpg highway-mpg    price  \
0                9.0       111.0    5000.0       21          27  13495.0   
1                9.0       111.0    5000.0       21          27  16500.0   
2                9.0       154.0    5000.0       19          26  16500.0   
3               10.0       102.0    5500.0       24          30  13950.0   
4                8.0       115.0    5500.0       18          22  17450.0   

  city-L/100km  horsepower-binned  diesel  gas  
0    11.190476             Medium       0    1  
1    11.190476             Medium       0    1  
2    12.368421             Medium       0    1  
3     9.791667             Medium       0    1  
4    13.055556             Medium       0    1  

[5 rows x 29 columns]

1. Linear Regression and Multiple Linear Regression
Linear Regression
One example of a Data Model that we will be using is

Simple Linear Regression.
Simple Linear Regression is a method to help us understand the relationship between two variables:

𝑌:𝑅𝑒𝑠𝑝𝑜𝑛𝑠𝑒 𝑉𝑎𝑟𝑖𝑎𝑏𝑙𝑒
𝑋:𝑃𝑟𝑒𝑑𝑖𝑐𝑡𝑜𝑟 𝑉𝑎𝑟𝑖𝑎𝑏𝑙𝑒𝑠
Linear function:
𝑌ℎ𝑎𝑡=𝑎+𝑏𝑋

Lets load the modules for linear regression

In [4]:
from sklearn.linear_model import LinearRegression

Next: 
Create the linear regression object

In [5]:
lm = LinearRegression()
lm

LinearRegression()

How could Highway-mpg help us predict car price?
For this example, we want to look at how highway-mpg can help us predict car price. Using simple linear regression, we will create a linear function with "highway-mpg" as the predictor variable and the "price" as the response variable.

In [6]:
X = cars[['highway-mpg']]
Y = cars[['price']]

Next:
Fit the linear model using highway-mpg.

In [7]:
lm.fit(X,Y)

LinearRegression()

Next:
We can output a prediction

In [8]:
Yhat = lm.predict(X)
Yhat[0:5]

array([[16236.50464347],
       [16236.50464347],
       [17058.23802179],
       [13771.3045085 ],
       [20345.17153508]])

Next
Find the value of the intercept (a)?

In [9]:
lm.intercept_

array([38423.30585816])

Next:
What is the value of the Slope (b)?

In [10]:
lm.coef_

array([[-821.73337832]])

Plugging in the actual values we get:

price = 38423.31 - 821.73 x highway-mpg

Next:
Train the model using 'engine-size' as the independent variable and 'price' as the dependent variable?

In [11]:
lm1 = LinearRegression()
X = cars[['engine-size']]
Y = cars[['price']]

lm1.fit(X, Y)


LinearRegression()

Next:
Find the slope and intercept of the model?

In [12]:
lm1.coef_

array([[166.86001569]])

In [13]:
lm1.intercept_

array([-7963.33890628])

Multiple Linear Regression
If we want to use more variables in our model to predict car price, we can use Multiple Linear Regression. 

From the previous section we know that other good predictors of price could be:

Horsepower
Curb-weight
Engine-size
Highway-mpg
Let's develop a model using these variables as the predictor variables.

In [14]:
Z3 = cars[['horsepower', 'curb-weight', 'engine-size', 'highway-mpg']]

Fit the linear model using the four above-mentioned variables.

In [15]:
lm3 = LinearRegression()
lm3.fit(Z3, cars[['price']])
lm3

LinearRegression()

What is the value of the intercept(a)?

In [16]:
lm3.intercept_

array([-15806.62462633])

What are the values of the coefficients (b1, b2, b3, b4)?

In [17]:
lm3.coef_

array([[53.49574423,  4.70770099, 81.53026382, 36.05748882]])

As we saw above, we should get a final linear function with the structure:

𝑌ℎ𝑎𝑡=𝑎+𝑏1𝑋1+𝑏2𝑋2+𝑏3𝑋3+𝑏4𝑋4
 
What is the linear function we get in this example?

Price = -15678.742628061467 + 52.65851272 x horsepower + 4.69878948 x curb-weight + 81.95906216 x engine-size + 33.58258185 x highway-mpg

Next:
Create and train a Multiple Linear Regression model "lm4" where the response variable is price, and the predictor variable is 'normalized-losses' and 'highway-mpg'.

In [18]:
lm4 = LinearRegression()
Z4 = cars[['normalized-losses', 'highway-mpg']]
Y = cars[['price']]

lm4.fit(Z4, Y)

LinearRegression()

In [19]:
Next
Find the coefficient of the model

SyntaxError: invalid syntax (<ipython-input-19-35708ebebd0d>, line 2)

In [ ]:
lm4.coef_

Model Evaluation using Visualization
Now that we've developed some models, how do we evaluate our models and how do we choose the best one? One way to do this is by using visualization.

import the visualization package: seaborn

In [ ]:
import seaborn as sns
%matplotlib inline

Regression Plot
When it comes to simple linear regression, an excellent way to visualize the fit of our model is by using regression plots.

Let's visualize Horsepower as potential predictor variable of price:

In [ ]:
width = 8
height = 6 
plt.figure(figsize=(width, height))
sns.regplot(x='highway-mpg', y='price', data=cars)
plt.ylim(0,)

We can see from this plot that price is negatively correlated to highway-mpg, since the regression slope is negative. One thing to keep in mind when looking at a regression plot is to pay attention to how scattered the data points are around the regression line. This will give you a good indication of the variance of the data, and whether a linear model would be the best fit or not. If the data is too far off from the line, this linear model might not be the best model for this data. 

Let's compare this plot to the regression plot of "peak-rpm".

In [ ]:
plt.figure(figsize=(width, height))
sns.regplot(x='peak-rpm', y='price', data=cars)
plt.ylim(0,)

Comparing the regression plot of "peak-rpm" and "highway-mpg" we see that the points for "highway-mpg" are much closer to the generated line and on the average decrease. The points for "peak-rpm" have more spread around the predicted line, and it is much harder to determine if the points are decreasing or increasing as the "highway-mpg" increases.

Given the regression plots above is "peak-rpm" or "highway-mpg" more strongly correlated with "price". Use the method ".corr()" to verify your answer.

In [ ]:
cars[['peak-rpm', 'highway-mpg', 'price']].corr()

Residual Plot
A good way to visualize the variance of the data is to use a residual plot.

- If the points in a residual plot are randomly spread out around the x-axis, then a linear model is appropriate for the data. Why is that? Randomly spread out residuals means that the variance is constant, and thus the linear model is a good fit for this data.

Let's find out what the residual plot looks like for between ['highway-mpg'] and ['price']

In [ ]:
width = 8
height = 6 
plt.figure(figsize=(width, height))
sns.residplot(x='highway-mpg', y='price', data=cars)
plt.show()

What is this plot telling us?

We can see from this residual plot that the residuals are not randomly spread around the x-axis, which leads us to believe that maybe a non-linear model is more appropriate for this data.

Multiple Linear Regression
How do we visualize a model for Multiple Linear Regression? This gets a bit more complicated because you can't visualize it with regression or residual plot.

One way to look at the fit of the model is by looking at the distribution plot: We can look at the distribution of the fitted values that result from the model and compare it to the distribution of the actual values.


First lets make a prediction using lm4 and Z4 = cars[['peak-rpm', 'highway-mpg']]

In [ ]:
Y4_hat = lm4.predict(Z4)

In [ ]:
width = 6
height = 5

plt.figure(figsize=(width, height))

ax4 = sns.distplot(cars['price'], hist=False, color='r', label='Actual Price')
sns.distplot(Y4_hat, hist=False, color='g', label='Predict Price', ax=ax4)

plt.xlabel('Price in dollars')
plt.ylabel('Fractions of Cars ')
plt.show()
plt.close()

We can see that the fitted values are reasonably close to the actual values, since the two distributions overlap a bit. However, there is definitely some room for improvement.

Next lets make a prediction using lm3 and Z3 = cars[['horsepower', 'curb-weight', 'engine-size', 'highway-mpg']]

In [ ]:
Y3_hat = lm3.predict(Z3)

In [ ]:
width = 8 
height = 6
plt.figure(figsize=(width, height))

ax3 = sns.distplot(cars['price'], hist=False, color='r', label='Actual Price')
sns.distplot(Y3_hat, hist=False, color='b', label='Predict Price', ax=ax3)

plt.title('Actual Price vs Predict Price for Cars')
plt.xlabel('Price in dollars')
plt.ylabel('Proportions of cars')
plt.show()
plt.close()

We can see that the fitted values has improved compared to the first model. 

Part 3: Polynomial Regression and Pipelines

Polynomial regression is a particular case of the general linear regression model or multiple linear regression models.

We get non-linear relationships by squaring or setting higher-order terms of the predictor variables.

There are different orders of polynomial regression:

Quadratic - 2nd order
𝑌ℎ𝑎𝑡=𝑎+𝑏1𝑋^2+𝑏2𝑋^2
 
Cubic - 3rd order
𝑌ℎ𝑎𝑡=𝑎+𝑏1𝑋^2+𝑏2𝑋^2+𝑏3𝑋^3

Higher order:
𝑌=𝑎+𝑏1𝑋^2+𝑏2𝑋^2+𝑏3𝑋^3....

We saw earlier that a linear model did not provide the best fit while using highway-mpg as the predictor variable. Let's see if we can try fitting a polynomial model to the data instead.

We will use the following function to plot the data:

In [ ]:
def Plotpolly(model, x, y, x_label):
    
    x_new = np.linspace(15, 55, 100)
    y_predict = model(x_new)
    
    plt.plot(x, y, '.', x_new, y_predict, '-')
    plt.title('Polynomial fit with matplotlib for price & ', x_label)
    

    